// @cds/core/tag/tag.element.js
/*
 * Copyright (c) 2016-2021 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { assignSlotNames, baseStyles, CdsBaseButton, id, property, syncProps } from '@cds/core/internal';
import { html } from 'lit-element';
import { styles } from './tag.element.css.js';
/**
 * Tags show concise metadata in a compact format.
 * Tags are visually styled to differentiate them from buttons.
 *
 * ```typescript
 * import '@cds/core/tag/register.js';
 * ```
 *
 * ```html
 * <cds-tag status="info">Info</cds-tag>
 * ```
 *
 * @element cds-tag
 * @slot - Content slot for inside the tag
 * @cssprop --background
 * @cssprop --border-color
 * @cssprop --border-radius
 * @cssprop --border-width
 * @cssprop --color
 * @cssprop --font-size
 * @cssprop --padding
 * @cssprop --size
 */
export class CdsTag extends CdsBaseButton {
    constructor() {
        super(...arguments);
        /**
         * @type {neutral | info | success | warning | danger}
         * Sets the color of the tag (and badge if present) from the following predefined list of statuses:
         */
        this.status = 'neutral';
        /** If present, shows a close icon to one side of the tag.
         *  Note that applications must wire up the action to remove the tag on click and also
         *  that you cannot have a clickable AND closable tag. The closable attribute-property
         *  shows the close icon. What happens when the tag is clicked is for application developers
         *  to decide.
         *  If closable is present, the tag will be considered "clickable".
         */
        this.closable = false;
    }
    static get styles() {
        return [baseStyles, styles];
    }
    connectedCallback() {
        super.connectedCallback();
        assignSlotNames([this.icon, 'tag-icon'], [this.badge, 'tag-badge']);
    }
    updated(props) {
        super.updated(props);
        if (props.has('closable') && props.get('closable') === true) {
            this.readonly = false;
        }
        if (this.badge) {
            syncProps(this.badge, this, { status: props.has('status'), color: props.has('color') });
        }
    }
    render() {
        return html `
      <div
        class="private-host"
        role="group"
        aria-labelledby="${this.groupLabelId}"
        cds-layout="horizontal align:vertical-center"
      >
        <slot name="tag-icon"></slot>
        <span id="${this.groupLabelId}" class="tag-content" cds-text="lhe"><slot></slot></span>
        <slot name="tag-badge"></slot>
        ${this.closable ? html `<cds-icon shape="times"></cds-icon>` : html ``}
      </div>
    `;
    }
}
__decorate([
    property({ type: String })
], CdsTag.prototype, "status", void 0);
__decorate([
    property({ type: String })
], CdsTag.prototype, "color", void 0);
__decorate([
    property({ type: Boolean })
], CdsTag.prototype, "closable", void 0);
__decorate([
    id()
], CdsTag.prototype, "groupLabelId", void 0);
