// @cds/core/internal/base/button.base.js
/*
 * Copyright (c) 2016-2021 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { __decorate } from "tslib";
import { html, LitElement } from 'lit-element';
import { ifDefined } from 'lit-html/directives/if-defined.js';
import { property, internalProperty } from '../decorators/property.js';
import { querySlot } from '../decorators/query-slot.js';
import { onAnyKey } from '../utils/keycodes.js';
import { stopEvent } from './../utils/events.js';
// @dynamic
export class CdsBaseButton extends LitElement {
    constructor() {
        super(...arguments);
        this.readonly = false;
        this.disabled = false;
        this.ariaDisabled = 'false';
        this.focused = false;
        this.active = false;
        this.role = 'button';
        this.isAnchor = false;
    }
    render() {
        return html ` <slot></slot> `;
    }
    connectedCallback() {
        super.connectedCallback();
        this.tabIndex = 0; // initialize immediately so button can be focused synchronously
    }
    updated(props) {
        super.updated(props);
        this.updateButtonAttributes(props);
    }
    /** This mimics the mouse-click visual behavior for keyboard only users and screen readers.
     * Browsers do not apply the CSS psuedo-selector :active in those instances. So we need this
     * for our :active styles to show.
     *
     * Make sure to update a component's CSS to account for the presence of the [_active] attribute
     * in all instance where :active is defined.
     *
     * @private
     */
    showClick() {
        this.active = true;
        const clickTimer = setTimeout(() => {
            this.active = false;
            clearTimeout(clickTimer);
        }, 300);
    }
    setupNativeButtonBehavior(readonly) {
        if (readonly) {
            this.removeEventListener('click', this.triggerNativeButtonBehavior);
            this.removeEventListener('keydown', this.emulateKeyBoardEventBehavior);
        }
        else {
            this.addEventListener('click', this.triggerNativeButtonBehavior);
            this.addEventListener('keydown', this.emulateKeyBoardEventBehavior);
        }
    }
    /**
     * We have to append a hidden button outside the web component in the light DOM
     * This allows us to trigger native submit events within a form element.
     */
    triggerNativeButtonBehavior(event) {
        if (!this.readonly) {
            if (this.disabled) {
                stopEvent(event);
            }
            else if (!event.defaultPrevented) {
                const buttonTemplate = html `<button
          class="cds-hidden-button"
          aria-hidden="true"
          ?disabled="${this.disabled}"
          tabindex="-1"
          style="display: none !important"
          value="${ifDefined(this.value)}"
          name="${ifDefined(this.name)}"
          type="${ifDefined(this.type)}"
        ></button>`;
                this.showClick();
                this.appendChild(buttonTemplate.getTemplateElement().content.cloneNode(true));
                const button = this.querySelector('.cds-hidden-button');
                button.dispatchEvent(new MouseEvent('click', { relatedTarget: this, composed: true }));
                button.remove();
            }
        }
    }
    emulateKeyBoardEventBehavior(evt) {
        onAnyKey(['enter', 'space'], evt, () => {
            this.click();
            stopEvent(evt);
        });
    }
    updateButtonAttributes(props) {
        var _a;
        this.isAnchor = ((_a = this.parentElement) === null || _a === void 0 ? void 0 : _a.tagName) === 'A';
        if (this.isAnchor && this.parentElement) {
            this.parentElement.style.lineHeight = '0';
            this.parentElement.style.textDecoration = 'none'; // fixes issue when style is applied to text node
        }
        this.readonly = this.readonly || this.isAnchor;
        if (this.readonly) {
            this.role = null;
            this.tabIndexAttr = null;
            this.ariaDisabled = null;
        }
        else {
            this.role = 'button';
            this.tabIndexAttr = this.disabled ? -1 : 0;
            this.ariaDisabled = this.disabled ? 'true' : 'false';
        }
        if (props.has('readonly')) {
            this.setupNativeButtonBehavior(this.readonly);
        }
    }
}
__decorate([
    property({ type: Boolean })
], CdsBaseButton.prototype, "readonly", void 0);
__decorate([
    property({ type: String })
], CdsBaseButton.prototype, "type", void 0);
__decorate([
    property({ type: String })
], CdsBaseButton.prototype, "name", void 0);
__decorate([
    property({ type: String })
], CdsBaseButton.prototype, "value", void 0);
__decorate([
    property({ type: Boolean })
], CdsBaseButton.prototype, "disabled", void 0);
__decorate([
    internalProperty({ type: String, attribute: 'aria-disabled', reflect: true })
], CdsBaseButton.prototype, "ariaDisabled", void 0);
__decorate([
    internalProperty({ type: Number, attribute: 'tabindex', reflect: true })
], CdsBaseButton.prototype, "tabIndexAttr", void 0);
__decorate([
    internalProperty({ type: Boolean, reflect: true })
], CdsBaseButton.prototype, "focused", void 0);
__decorate([
    internalProperty({ type: Boolean, reflect: true })
], CdsBaseButton.prototype, "active", void 0);
__decorate([
    internalProperty({ type: String, reflect: true, attribute: 'role' })
], CdsBaseButton.prototype, "role", void 0);
__decorate([
    internalProperty({ type: Boolean, reflect: true })
], CdsBaseButton.prototype, "isAnchor", void 0);
__decorate([
    querySlot('cds-icon')
], CdsBaseButton.prototype, "icon", void 0);
__decorate([
    querySlot('cds-badge')
], CdsBaseButton.prototype, "badge", void 0);
